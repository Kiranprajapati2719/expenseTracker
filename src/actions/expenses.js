import database from '../firebase/firebase';

//add expense
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { description, note, amount, createdAt };
    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

//remove expense

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

//edit expense

export const editExpense = (id, update) => ({
  type: 'EDIT_EXPENSE',
  id,
  update
}); 

//SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = (expenseData = {}) => {
  return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = [];
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      })
      dispatch(setExpenses(expenses));
    });
  }
};