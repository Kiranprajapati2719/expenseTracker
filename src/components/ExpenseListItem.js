import React from 'react'
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import { Container, Segment, Grid } from 'semantic-ui-react';

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <Container>
    <Segment>
      <Grid container textAlign='center'>
        <Grid.Column floated='left' textAlign='left' width={12}>
          <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
          </Link>
        </Grid.Column>
        <Grid.Column floated='right' width={4}>
          <h3>{numeral(amount / 100).format('$, 0.00')}</h3>
          <p className='normalFont'>{moment(createdAt).format('MMM Do, YYYY')}</p>
        </Grid.Column>
      </Grid>
    </Segment>
  </Container>
);
export default ExpenseListItem; 