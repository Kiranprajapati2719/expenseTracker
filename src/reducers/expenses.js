
//Expense Reducer

const defaultExpenses = [];

export default (state = defaultExpenses, action) => {
  switch(action.type){
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense
      ];
    case 'REMOVE_EXPENSE':
      //remove selected id
      return state.filter(({id})=> id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map((expense)=>{
        if(expense.id === action.id){
          return {
            //override with existing
            ...expense,
            ...action.update
          }; 
        }else{
          return expense;
        }
      });
    case 'SET_EXPENSES' : 
      return action.expenses;
    default:
      return state;
  }
};
