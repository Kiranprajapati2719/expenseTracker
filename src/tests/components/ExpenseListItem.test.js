import ExpenseListItem from '../../components/ExpenseListItem';
import {shallow} from 'enzyme'; 
import React from 'react';
import expenses from '../fixtures/expenses';

test('Should render expense list item correctly', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
  expect(wrapper).toMatchSnapshot();
}); 
    