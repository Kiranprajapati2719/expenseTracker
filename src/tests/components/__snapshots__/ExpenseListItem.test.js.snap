// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expense list item correctly 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Rent
    </h3>
  </Link>
  <p>
    Amount: 
    $90.00
    <br />
    Created: 
    1/1/1970
    <br />
  </p>
</div>
`;
